### ISSUE TITLE
Travis job does not complete (minion is waiting for a leader; leader is waiting for the minion)

### ISSUE THREAD

**Issue by gajus** [semantic-release/semantic-release#274]
This is the leader:

https://travis-ci.org/gajus/gitinfo/jobs/152546543

```
This is a leader
Leader waits for minions [u'92.2', u'92.3']...
Leader waits for minions [u'92.2', u'92.3']...
Leader waits for minions [u'92.2', u'92.3']...
Leader waits for minions [u'92.2', u'92.3']...
Leader waits for minions [u'92.2', u'92.3']...
Leader waits for minions [u'92.2', u'92.3']...
Leader waits for minions [u'92.2']...
Leader waits for minions [u'92.2']...
Leader waits for minions [u'92.2']...
Leader waits for minions [u'92.2']...
Leader waits for minions [u'92.2']...
```

It appears to be waiting for job #92.2.

Here is the job #92.2:

https://travis-ci.org/gajus/gitinfo/jobs/152546544

```
[travis-after-all] Job succeeded (8410eebc89b4c2d11691744e52da09fd89c48ac5fd599c7ef564c66bac130664)
[travis-after-all] Waiting for 92.1 and 92.3 to be done...
[travis-after-all] Waiting for 92.1 and 92.3 to be done...
[travis-after-all] Waiting for 92.1 to be done...
[travis-after-all] Waiting for 92.1 to be done...
[travis-after-all] Waiting for 92.1 to be done...
```

It appears to be waiting for the leader to complete.

This is my `.travis-ci.yml` configuration:

``` yaml
language: node_js
node_js:
    - 6
    - 5
    - 4
before_install:
    - npm install --global npm@3
before_script:
    - curl -Lo travis_after_all.py https://git.io/vLSON
script:
    - NODE_ENV=development npm run build
    - rm -fr ./dist
    - NODE_ENV=production npm run build
    - npm run lint
    - flow
    - nyc --silent npm run test
    - nyc report --reporter=text-lcov | coveralls
    - nyc check-coverage --lines 90
    - conventional-changelog-lint --from=HEAD~$(git --no-pager rev-list master..HEAD --count)
after_success:
    - python travis_after_all.py
    - export $(cat .to_export_back)
    - git config --global user.name "semantic-release (via TravisCI)"
    - git config --global user.email "semantic-release@travis"
    - semantic-release pre && npm publish && semantic-release post
notifications:
    email: false
sudo: false
```

What am I missing?

**Comment by gajus** [comment]
@boennemann any guidelines regarding this?

**Comment by vrtxf** [comment]
the solution is to restart the failing build (the stuck one)
`travis_after_all.py` is natively included in travis now afaik (it's still bugged anyway will work in 50% of the cases as intended)

**Comment by woshilapin** [comment]
@gajus I reached the same problem and found a solution.
TL;DR; remove the Python script `travis-after-all.py` from your `.travis.yml`.

I didn't have the problem until I updated `semantic-release` from 4.3.5 to version 6.3.1. And it seems that this new version is including a functionality similar to `travis_after_all.py`.  The dependency is  `travis-after-all` in semantic-release (see dependency [`@semantic-release/condition-travis`](https://github.com/semantic-release/condition-travis) which itself has the dependency [`travis-after-all`](https://www.npmjs.com/package/travis-after-all)).

What was happening was that the Python script got called on all the jobs but only the first job was proclaimed leader (which is what is intended). The other jobs are launching the other commands and are reaching `semantic-release pre`. And the first job that called `semantic-release pre` will launch the `travis-after-all` javascript script and will also be proclaimed leader.  Second one will endlessly wait for the first one, and first one will endlessly wait for the second: the Python script and the Javascript dependency are fighting to do the exact same thing.

The Python script is now redundant with the `semantic-release` dependency `travis-after-all`.  You can check in your build this information.  First of all, the format of the messages are different (`Leader waits for minions [u'92.2']...` vs `[travis-after-all] Waiting for 92.1 to be done...`). Second, in your build `#92.1`, line 1403, it's the Python build which is waiting.  However, in `#92.2`, line 1501, the Python script has identified the job as a minion and line 1509, `semantic-release` is launched, waiting for other jobs to finish.  Hope this helps.

**Comment by boennemann** [comment]
Thanks for the explaining this @woshilapin. My explanation for this was buried in the 6.0.0 release notes.

semantic-release 6.3.2 is now on latest. It depends on [travis-deploy-once](npm.im/travis-deploy-once) instead of travis_after_all, or travis-after-all. You have to remove the python script from the travis file to get this to work.

The CLI now sets this up correctly, too.
